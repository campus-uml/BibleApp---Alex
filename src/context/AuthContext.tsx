import { createContext, useContext, useEffect, useState } from "react";
import { supabase } from "..//constants/api";
import { User } from "@/types";

interface AuthContextType {
  user: User | null;
  logout: () => void;
}

const AuthContext = createContext<AuthContextType | null>(null);

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const [user, setUser] = useState<User | null>(null);

  useEffect(() => {
    const fetchUser = async () => {
      const { data } = await supabase.auth.getUser();
      if (data.user) setUser(data.user);
    };

    fetchUser();

    const { data: listener } = supabase.auth.onAuthStateChange((_event, session) => {
      setUser(session?.user || null);
    });

    return () => {
      listener?.subscription.unsubscribe();
    };
  }, []);

  const logout = async () => {
    await supabase.auth.signOut();
    setUser(null);
  };

  return (
    <AuthContext.Provider value={{ user, logout }}>
      {children}
    </AuthContext.Provider>
  );
};

// eslint-disable-next-line react-refresh/only-export-components
export const useAuth = () => {
  return useContext(AuthContext);
};
