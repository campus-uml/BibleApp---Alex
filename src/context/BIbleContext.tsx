import React, {
  createContext,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";
import {
  getBibles,
  getChapters,
  loadChapterVersesFromAPI,
  searchBible,
} from "../services/getData";
import { BibleBooks, Chapter, Verse } from "../types/index";
import { useSidebar } from "@/components/ui/sidebar";

const formatVerseText = (text: string) => {
  return text.toLowerCase().replace(/^([a-z])/, (match) => match.toUpperCase());
};

const isTitleOrHeading = (text: string) => {
  const titleKeywords = ["Capítulo", "Libro", "Título"];
  return titleKeywords.some((keyword) => text.includes(keyword));
};

interface BibleContextProps {
  bibleVerse: BibleBooks[];
  selectedBook: BibleBooks["id"];
  bibleVerseChapters: Chapter[];
  chapterVerses: Verse[];
  scrollAreaRef: React.RefObject<HTMLDivElement>;
  handleBook: (bookId: string) => void;
  loadChapterVerses: (chapterId: string) => void;
  searchBibleVerse: (query: string) => void;
  toggleSidebar: () => void;
  
}

const BibleContext = createContext<BibleContextProps | undefined>(undefined);

interface BibleProviderProps {
  children: React.ReactNode;
}

export const BibleProvider: React.FC<BibleProviderProps> = ({ children }) => {
  const [bibleVerse, setBibleVerse] = useState<BibleBooks[]>([]);
  const [selectedBook, setSelectedBook] = useState<BibleBooks["id"]>("GEN");
  const [bibleVerseChapters, setBibleVerseChapters] = useState<Chapter[]>([]);
  const [chapterVerses, setChapterVerses] = useState<Verse[]>([]);
  const scrollAreaRef = useRef<HTMLDivElement>(null);
  const { toggleSidebar } = useSidebar();

  useEffect(() => {
    fetchData();
  }, []);

  useEffect(() => {
    if (selectedBook) {
      loadChapters(selectedBook);
    }
  }, [selectedBook]);

  useEffect(() => {
    if (bibleVerseChapters.length > 0) {
      loadChapterVerses(bibleVerseChapters[0].id);
    }
  }, [bibleVerseChapters]);

  const loadChapterVerses = async (chapterId: string) => {
    try {
      const response = await loadChapterVersesFromAPI(chapterId);
      if (!response.data) {
        console.error(
          "Error: No se encontró contenido en los versículos",
          response
        );
        return;
      }

      const content = response.data.content.replace(/<\/?[^>]+(>|$)/g, "");
      const versePattern = /(\d+)([^0-9]+)/g;
      const versesArray: Verse[] = [];
      let match;

      while ((match = versePattern.exec(content)) !== null) {
        const verseNumber = match[1];
        const verseText = formatVerseText(match[2].trim());

        if (!isTitleOrHeading(verseText)) {
          versesArray.push({
            id: `${chapterId}-${verseNumber}`,
            reference: `Versículo ${verseNumber}`,
            text: verseText,
          });
        }
      }

      setChapterVerses(versesArray);
    } catch (error) {
      console.error("Error fetching chapter content:", error);
    }
  };

  const fetchData = async () => {
    try {
      const response = await getBibles();
      if (response && response.data) {
        setBibleVerse(response.data);
      } else {
        console.error("Unexpected API response structure:", response);
      }
    } catch (error) {
      console.error("Error fetching bibles:", error);
    }
  };

  const handleBook = (bookId: string) => {
    setSelectedBook(bookId);
    toggleSidebar();
  };

  const loadChapters = async (bookId: string) => {
    try {
      const response = await getChapters(bookId);
      if (response && response.data) {
        setBibleVerseChapters(response.data);
      } else {
        console.error("Unexpected API response structure:", response);
      }
    } catch (error) {
      console.error("Error fetching chapters:", error);
    }
  };

  const searchBibleVerse = async (query: string) => {
    try {
      const response = await searchBible(query);
      if (response && response.data) {
        setBibleVerse(response.data);
      } else {
        console.error("Unexpected API response structure:", response);
      }
    } catch (error) {
      console.error("Error fetching bible verse:", error);
    }
  };
  return (
    <BibleContext.Provider
      value={{
        bibleVerse,
        selectedBook,
        bibleVerseChapters,
        chapterVerses,
        handleBook,
        loadChapterVerses,
        scrollAreaRef,
        searchBibleVerse,
        toggleSidebar,
      }}
    >
      {children}
    </BibleContext.Provider>
  );
};

// eslint-disable-next-line react-refresh/only-export-components
export const useBible = (): BibleContextProps => {
  const context = useContext(BibleContext);
  if (!context) {
    throw new Error("useBible must be used within a BibleProvider");
  }
  return context;
};
