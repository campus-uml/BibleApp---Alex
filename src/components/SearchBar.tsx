import React, { useState } from "react";
import { useBible } from "../context/BIbleContext";
import { Button } from "./ui/button";

const SearchBar: React.FC = () => {
  const { searchBibleVerse } = useBible();
  const [query, setQuery] = useState("");

  const handleSearch = () => {
    if (query.trim() !== "") {
      searchBibleVerse(query);
    }
  };

  return (
    <div className="flex items-center space-x-2 ">
      <input
        type="text"
        value={query}
        onChange={(e) => setQuery(e.target.value)}
        placeholder="Buscar en la Biblia..."
        className="border rounded-lg p-2 w-full"
      />
      <Button onClick={handleSearch} className="p-2  rounded-lg transition duration-300">Buscar</Button>
    </div>
  );
};

export default SearchBar;
