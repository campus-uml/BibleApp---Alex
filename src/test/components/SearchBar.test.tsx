import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import type { Mock } from "vitest";
import SearchBar from "../../components/SearchBar";
import { useBible } from "../../context/BIbleContext";

vi.mock("../../context/BIbleContext");

describe("SearchBar", () => {
  it("llama a searchBibleVerse al enviar el formulario", async () => {
    const searchBibleVerse = vi.fn();
    (useBible as Mock).mockReturnValue({ searchBibleVerse });

    const { asFragment } = render(<SearchBar />);
    const input = screen.getByPlaceholderText(
      "Ejemplo: 'Dios', 'fe', 'esperanza'"
    );
    const button = screen.getByRole("button");

    fireEvent.change(input, { target: { value: "Dios" } });
    fireEvent.click(button);

    expect(searchBibleVerse).toHaveBeenCalledWith("Dios");
    expect(asFragment()).toMatchSnapshot();
  });

  it("deshabilita el input y el botÃ³n mientras se carga", async () => {
    const searchBibleVerse = vi.fn();
    (useBible as Mock).mockReturnValue({ searchBibleVerse });

    const { asFragment } = render(<SearchBar />);
    const input = screen.getByPlaceholderText(
      "Ejemplo: 'Dios', 'fe', 'esperanza'"
    );
    const button = screen.getByRole("button");

    fireEvent.change(input, { target: { value: "Dios" } });
    fireEvent.click(button);

    expect(input).toBeDisabled();
    expect(button).toBeDisabled();
    expect(asFragment()).toMatchSnapshot();
  });
});
