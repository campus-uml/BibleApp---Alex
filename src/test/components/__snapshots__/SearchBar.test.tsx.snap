// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > deshabilita el input y el bot√≥n mientras se carga 1`] = `
<DocumentFragment>
  <form
    class="flex items-center gap-2 w-full"
  >
    <input
      class="border rounded-lg p-2 flex-grow sm:max-w-[250px] w-4/5 dark:bg-slate-800"
      disabled=""
      placeholder="Ejemplo: 'Dios', 'fe', 'esperanza'"
      type="text"
      value="Dios"
    />
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow hover:bg-primary/90 h-9 p-2 rounded-lg text-sidebar-primary-foreground bg-sidebar-primary dark:bg-slate-50 dark:text-slate-900"
      disabled=""
      type="submit"
    >
      <svg
        class="lucide lucide-loader-circle w-5 h-5 animate-spin"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 12a9 9 0 1 1-6.219-8.56"
        />
      </svg>
    </button>
  </form>
</DocumentFragment>
`;

exports[`SearchBar > llama a searchBibleVerse al enviar el formulario 1`] = `
<DocumentFragment>
  <form
    class="flex items-center gap-2 w-full"
  >
    <input
      class="border rounded-lg p-2 flex-grow sm:max-w-[250px] w-4/5 dark:bg-slate-800"
      disabled=""
      placeholder="Ejemplo: 'Dios', 'fe', 'esperanza'"
      type="text"
      value="Dios"
    />
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow hover:bg-primary/90 h-9 p-2 rounded-lg text-sidebar-primary-foreground bg-sidebar-primary dark:bg-slate-50 dark:text-slate-900"
      disabled=""
      type="submit"
    >
      <svg
        class="lucide lucide-loader-circle w-5 h-5 animate-spin"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 12a9 9 0 1 1-6.219-8.56"
        />
      </svg>
    </button>
  </form>
</DocumentFragment>
`;
